### PR TITLE
feat: error messages when connecting to MSSQL

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

Add better error messages when connecting to MSSQL.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Screen Shot 2021-04-12 at 6 49 48 PM
Screen Shot 2021-04-12 at 6 49 40 PM
Screen Shot 2021-04-12 at 6 49 05 PM
Screen Shot 2021-04-12 at 6 40 59 PM

### TEST PLAN

Added unit tests and tested locally (see screenshots).

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: docs/src/pages/docs/Miscellaneous/issue_codes.mdx [hunk 1]
```diff
@@ -149,3 +149,11 @@ The password provided when connecting to a database is not valid.
...
 The user provided a password that is incorrect. Please check that the
 password is typed correctly.
+
+## Issue 1014
+
+```
+Either the username or the password used are incorrect.
+```
+
+Either the username provided does not exist or the password was written incorrectly. Please check that the username and were typed correctly.
```
Comment (AAfghahi) [inline docs/src/pages/docs/Miscellaneous/issue_codes.mdx:159]:
```suggestion
Either the username provided does not exist or the password was written incorrectly. Please check that the username and password were typed correctly.
```

File: superset/db_engine_specs/mssql.py [hunk 2]
```diff
@@ -15,15 +15,33 @@
...
 from superset.db_engine_specs.base import BaseEngineSpec, LimitMethod
+from superset.errors import SupersetErrorType
 from superset.utils import core as utils
 
 logger = logging.getLogger(__name__)
 
 
+# Regular expressions to catch custom errors
+TEST_CONNECTION_ACCESS_DENIED_REGEX = re.compile("Adaptive Server connection failed")
+TEXT_CONNECTION_INVALID_HOSTNAME_REGEX = re.compile(
```
Comment (AAfghahi) [inline superset/db_engine_specs/mssql.py:33]:
```suggestion
TEST_CONNECTION_INVALID_HOSTNAME_REGEX = re.compile(
```

File: superset/db_engine_specs/mssql.py [hunk 3]
```diff
@@ -46,6 +64,28 @@ class MssqlEngineSpec(BaseEngineSpec):
         "P1Y": "DATEADD(year, DATEDIFF(year, 0, {col}), 0)",
     }
 
+    custom_errors = {
+        TEST_CONNECTION_ACCESS_DENIED_REGEX: (
+            __('Either the username "%(username)s" or the password is incorrect.'),
+            SupersetErrorType.TEST_CONNECTION_ACCESS_DENIED_ERROR,
+        ),
+        TEXT_CONNECTION_INVALID_HOSTNAME_REGEX: (
```
Comment (AAfghahi) [inline superset/db_engine_specs/mssql.py:72]:
```suggestion
        TEST_CONNECTION_INVALID_HOSTNAME_REGEX: (
```

File: tests/db_engine_specs/postgres_tests.py [hunk 4]
```diff
@@ -274,7 +298,15 @@ def test_extract_errors(self):
                     "and can't be reached on port 12345"
```
Comment (AAfghahi) [inline tests/db_engine_specs/postgres_tests.py:298]: I think example.com should be "example.com" for some reason I can't add this as a commit suggestion.

File: tests/db_engine_specs/postgres_tests.py [hunk 5]
```diff
@@ -254,7 +273,12 @@ def test_extract_errors(self):
                 error_type=SupersetErrorType.TEST_CONNECTION_PORT_CLOSED_ERROR,
                 message="Port 12345 on hostname localhost refused the connection.",
                 level=ErrorLevel.ERROR,
```
Comment (AAfghahi) [inline tests/db_engine_specs/postgres_tests.py:275]: same here "localhost"

File: superset/db_engine_specs/mssql.py (hunk 2 above)
Comment (betodealmeida) [reply superset/db_engine_specs/mssql.py:33]: Ha, thanks!

File: tests/db_engine_specs/postgres_tests.py (hunk 4 above)
Comment (betodealmeida) [reply tests/db_engine_specs/postgres_tests.py:298]: +1

File: superset/db_engine_specs/mssql.py [hunk 6]
```diff
@@ -15,15 +15,33 @@
...
+# Regular expressions to catch custom errors
+TEST_CONNECTION_ACCESS_DENIED_REGEX = re.compile("Adaptive Server connection failed")
+TEST_CONNECTION_INVALID_HOSTNAME_REGEX = re.compile(
+    r"Adaptive Server is unavailable or does not exist \((?P<hostname>.*?)\)"
+    "(?!.*Net-Lib error).*$"
+)
+TEST_CONNECTION_PORT_CLOSED_REGEX = re.compile(
+    r"Net-Lib error during Connection refused \(61\)"
+)
+TEST_CONNECTION_HOST_DOWN_REGEX = re.compile(
```
Comment (eschutho) [inline superset/db_engine_specs/mssql.py:40]: are we only going to use these errors for test connection, or will we be using them for database creation as well?

Comment (betodealmeida) [reply superset/db_engine_specs/mssql.py:40]: We can use them for both. Anytime we have a `DBAPI` exception we can call `extract_errors` to get these errors.

**Review by hughhhh — APPROVED** [review]

**Pull request merged**